staging/src/kubevirt.io/api/migrations/v1alpha1/types.go

```diff
@@ -0,0 +1,196 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package v1alpha1
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	k6tv1 "kubevirt.io/api/core/v1"
+)
+
+// MigrationPolicy holds migration policy (i.e. configurations) to apply to a VM or group of VMs
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+// +genclient:nonNamespaced
+type MigrationPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              MigrationPolicySpec `json:"spec" valid:"required"`
+	// +nullable
+	Status MigrationPolicyStatus `json:"status,omitempty"`
+}
+
+type MigrationPolicySpec struct {
+	Selectors *Selectors `json:"selectors"`
+
+	//+optional
+	AllowAutoConverge *bool `json:"allowAutoConverge,omitempty"`
+	//+optional
+	BandwidthPerMigration *resource.Quantity `json:"bandwidthPerMigration,omitempty"`
+	//+optional
+	CompletionTimeoutPerGiB *int64 `json:"completionTimeoutPerGiB,omitempty"`
+	//+optional
+	AllowPostCopy *bool `json:"allowPostCopy,omitempty"`
+	//+optional
+	DisableTLS *bool `json:"disableTLS,omitempty"`
+}
+
+type Selectors struct {
+	//+optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+	//+optional
+	VirtualMachineInstanceSelector *metav1.LabelSelector `json:"virtualMachineInstanceSelector,omitempty"`
+}
+
+type MigrationPolicyStatus struct {
+}
+
+// MigrationPolicyList is a list of MigrationPolicy
+//
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type MigrationPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	// +listType=atomic
+	Items []MigrationPolicy `json:"items"`
+}
+
+// GetMigrationConfByPolicy returns a new migration configuration. The new configuration attributes will be overridden
+// by the migration policy if the specified attributes were defined for this policy. Otherwise they wouldn't change.
+// The boolean returned value indicates if any changes were made to the configurations.
+func (m *MigrationPolicy) GetMigrationConfByPolicy(clusterMigrationConfigurations *k6tv1.MigrationConfiguration) (changed bool, err error) {
+	policySpec := m.Spec
+	changed = false
```

Comment:
did we agree on merging the global kubevirt CR values with the matched migration policies? 

I thought you either get the global policy entirely or get a migration policy entirely, not some combination of the two.

Reply:
Well, I'm not sure really. The latest working model of this design that I had in my head had every MigrationPolicy working as a completely standalone config.

@vladikr do you have any opinions here? What's should be the relationship between a MigrationPolicy and the global migration configuration in the KubeVirt CR? I had been working with assumption that it's an all or nothing scenario, where you either match a MigrationPolicy, and get exactly what's in there, or use the global configuration and get exactly what's in the global config... but not a combination of the two.

thoughts?

Reply:
> I had been working with assumption that it's an all or nothing scenario, where you either match a MigrationPolicy, and get exactly what's in there

The question, as I see it, is what does it mean to `get exactly what's in there` where some of the values are missing. Or in other words - what are the default values? The same default values we use for KubevirtCR?

I think that an admin that sets KubevirtCR's values treat them as default values for its cluster. It might be weird for an administrator to set these default values in KubevirtCR then see other defaults in an applied policy with missing values.

Reply:
> @vladikr do you have any opinions here? What's should be the relationship between a MigrationPolicy and the global migration configuration in the KubeVirt CR? I had been working with assumption that it's an all or nothing scenario, where you either match a MigrationPolicy, and get exactly what's in there, or use the global configuration and get exactly what's in the global config... but not a combination of the two.
>
> thoughts?

It's a good point. I do see a value in having the migration configuration in KubeVirt CR act as a default. We have a lot of tunables in the CR, some are not even exposed in the policy.
For example, the migration policy would reference only post-copy migration, but would not express anything about progress timeout, in this case, I'd expect a default value to be used.

My only concern is that with this PR we will add the values of the migration policy to the VMI status, but we will not store the default values from CR. 
This may be a bit surprsing.

Reply:
Alrighty, as long as we document and are consistent with the kubevirt CR being the default, this works for me.